### PR TITLE
flake

### PR DESCRIPTION
**Pull request by michaelsdr**

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -75,12 +75,24 @@ To see how a Momentum ResNet can be created using a ResNet, you can run:
 .. code:: python
 
    >>> import torch
-   >>> from momentumnet import transform
+   >>> from momentumnet import transform_to_momentumnet
    >>> from torchvision.models import resnet101
-   >>> mresnet101 = transform(resnet101(), gamma=0.99)
+   >>> mresnet101 = transform_to_momentumnet(resnet101(), gamma=0.99, use_backprop=False)
 
 This initiates a Momentum ResNet with weights of a pretrained Resnet-101 on ImageNet.
```
Comment (pierreablin) [inline README.rst:82]: need to specify pretrained=True in `resnet101`

File: README.rst [hunk 2]
```diff
@@ -75,12 +75,24 @@ To see how a Momentum ResNet can be created using a ResNet, you can run:
 .. code:: python
 
    >>> import torch
-   >>> from momentumnet import transform
+   >>> from momentumnet import transform_to_momentumnet
    >>> from torchvision.models import resnet101
-   >>> mresnet101 = transform(resnet101(), gamma=0.99)
+   >>> mresnet101 = transform_to_momentumnet(resnet101(), gamma=0.99, use_backprop=False)
```
Comment (pierreablin) [inline README.rst:80]: split it in two lines: `resnet = resnet101(pretrained=True)`, then mresnet = ...

File: doc/index.rst [hunk 3]
```diff
@@ -52,12 +52,24 @@ To see how a Momentum ResNet can be created using a ResNet, you can run:
 .. code:: python
 
    >>> import torch
-   >>> from momentumnet import transform
+   >>> from momentumnet import transform_to_momentumnet
    >>> from torchvision.models import resnet101
-   >>> mresnet101 = transform(resnet101(), gamma=0.99)
+   >>> mresnet101 = transform_to_momentumnet(resnet101(), gamma=0.99, use_backprop=False)
```
Comment (pierreablin) [inline doc/index.rst:57]: same here, split in ttwo lines

File: momentumnet/models.py [hunk 4]
```diff
@@ -583,7 +583,9 @@ def ResNet18(num_classes=1000, bn=True):
     return ResNet(BasicBlock, [2, 2, 2, 2], num_classes, norm_layer=norm_layer)
 
 
-def mResNet18(num_classes=1000, init_speed=0, gamma=0.9, use_backprop=False, bn=True):
+def mResNet18(
```
Comment (pierreablin) [inline momentumnet/models.py:586]: do we still need this file?

File: momentumnet/plot_memory.py [hunk 5]
```diff
@@ -12,7 +12,7 @@
```
Comment (pierreablin) [inline momentumnet/plot_memory.py:12]: why the new file address?

Comment (pierreablin) [reply momentumnet/plot_memory.py:12]: should not be at the root

File: momentumnet/resnet_to_momentumnet.py [hunk 6]
```diff
@@ -33,39 +39,36 @@ def transform(model, residual_layers=['layer1', 'layer2', 'layer3', 'layer4'],
...
-        splitted_key = residual_layer.split('.')
+        splitted_key = residual_layer.split(".")
         parent_module = momnet
         for i, key in enumerate(splitted_key):
             module = parent_module._modules[key]
             if i < len(splitted_key) - 1:
                 parent_module = module
         if not keep_first_layer:
-            momentumnet = nn.Sequential(MomentumNetTransform(module, gamma=gamma, use_backprop=use_backprop))
+            momentumnet = nn.Sequential(
```
Comment (pierreablin) [inline momentumnet/resnet_to_momentumnet.py:49]: why do we need a nn.Sequential here?

File: momentumnet/resnet_to_momentumnet.py [hunk 7]
```diff
@@ -33,39 +39,36 @@ def transform(model, residual_layers=['layer1', 'layer2', 'layer3', 'layer4'],
...
+                MomentumNetTransform(
```
Comment (pierreablin) [inline momentumnet/resnet_to_momentumnet.py:50]: same here

File: momentumnet/resnet_to_momentumnet.py [hunk 8]
```diff
@@ -33,39 +39,36 @@ def transform(model, residual_layers=['layer1', 'layer2', 'layer3', 'layer4'],
...
+                    )
+                )
+            )
             momentumnet = nn.Sequential(module[0], momentumnet)
         setattr(parent_module, key, momentumnet)
     return momnet
 
 
-
 def transform_minus_id(layer):
```
Comment (pierreablin) [inline momentumnet/resnet_to_momentumnet.py:67]: do we need this function ?

File: momentumnet/models.py (hunk 4 above)
Comment (michaelsdr) [reply momentumnet/models.py:586]: yes it is uses in the Cifar 10 Trainer

File: momentumnet/resnet_to_momentumnet.py (hunk 6 above)
Comment (michaelsdr) [reply momentumnet/resnet_to_momentumnet.py:49]: I found it useful in some cases

File: momentumnet/resnet_to_momentumnet.py (hunk 7 above)
Comment (michaelsdr) [reply momentumnet/resnet_to_momentumnet.py:50]: same

File: momentumnet/resnet_to_momentumnet.py (hunk 8 above)
Comment (michaelsdr) [reply momentumnet/resnet_to_momentumnet.py:67]: no thks

File: momentumnet/plot_memory.py (hunk 5 above)
Comment (michaelsdr) [reply momentumnet/plot_memory.py:12]: it takes a long time for this example to run with circle ci so i moved it for now

**Comment by pierreablin** [comment]
TODO list:

- merge MomentumNetTransform and MomentumNet, needs to only have one remaining class.
- add proper docstrings for MomentumNet with all arguments mentioned and explained, and the same example as in the doc/index.rst
- add a warning when gamma is too small
- change the 'skip_connection' argument to another term
- add more explanation about the parameters in the docstring
- add reference to the paper in the docstring
- replace 'attributes' to "parameters" in MomentumNet docstring
- fix userwarning in exact_rep_pytorch

In the doc:

- remove anaconda mention
- Explain a bit more the class MomentumNet in the index.rst file, and the example (mention that it is 10 layers, etc..)
- Say that you can transform a resnet in two lines of code
- find a way to remove all "methods" in the momentum net class in the doc


Examples:

- No savefig
- Simple classif example with resnet + MLP ? eg on digits dataset
- Example on two nested rings separation

Then, make a new release !

**Comment by michaelsdr** [comment]
@pierreablin is it ok to merge?

File: momentumnet/resnet_to_momentumnet.py [hunk 9]
```diff
@@ -21,51 +28,71 @@ def transform(model, residual_layers=['layer1', 'layer2', 'layer3', 'layer4'],
         The resnet one desires to turn into a Momentum ResNet.
     residual_layers : a list of strings
```
[the remark this answers is not in the document]
Comment (pierreablin) [reply momentumnet/resnet_to_momentumnet.py:29]: In the examples, you could build for instance a fake resnet with different layer names to clealry show what the argument 'residual_layers' is used for

**Comment by pierreablin** [comment]
Still some work to do !

File: MANIFEST.in [hunk 10]
```diff
@@ -3,10 +3,10 @@ include LICENSE
 include momentumnet/__init__.py
 include *.png
 include requirements.txt
-
+include .DS_Store
```
[the remark this answers is not in the document]
Comment (michaelsdr) [reply MANIFEST.in:6]: so i exluded it (?)

File: momentumnet/resnet_to_momentumnet.py [hunk 11]
```diff
@@ -21,51 +28,71 @@ def transform(model, residual_layers=['layer1', 'layer2', 'layer3', 'layer4'],
...
         The momentum term for the Momentum ResNet.
     use_backprop : bool (default: False)
         If True then the Momentum ResNet has a smaller memory footprint.
+    is_residual : bool (default: False)
+        If True then the forward rule is x + f(x)
 
-    Return
-    ------
+    Returns
+    -------
```
[the remark this answers is not in the document]
Comment (michaelsdr) [reply momentumnet/resnet_to_momentumnet.py:43]: i like mresnet :)

File: momentumnet/tests/test_momentum_net.py [hunk 12]
```diff
@@ -50,14 +50,14 @@ def test_outputs_memory(init_speed):
         init_function = None
     mom_no_backprop = MomentumNet(
         functions,
-        gamma=0.9,
+        gamma=0.999,
```
[the remark this answers is not in the document]
Comment (michaelsdr) [reply momentumnet/tests/test_momentum_net.py:53]: no i changed it to 0.9

File: momentumnet/momentum_net.py [hunk 13]
```diff
@@ -247,21 +329,31 @@ def __init__(
...
+                init_speed,
+                init_function,
+                is_residual,
             )
         self.use_backprop = use_backprop
         self.gamma = gamma
         self.init_speed = init_speed
 
-    def forward(self, x, *args):
+    def forward(self, x, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (michaelsdr) [reply momentumnet/momentum_net.py:357]: i need it

**Comment by michaelsdr** [comment]
What remains 
TODO list:
* add a warning when gamma is too small
* fix userwarning in exact_rep_pytorch
In the doc:
* find a way to remove all "methods" in the momentum net class in the doc
Examples:
* Simple classif example with resnet + MLP ? eg on digits dataset
* Example on two nested rings separation

**Pull request merged**